Pioneer DDJ-200: fix for hotcue issues, rewrite

absorb-it opened the pull request:
The current implementation for this controller has some issues with the hotcue buttons not working as expected.

This is a rewrite of this controller implementation which fixes these issues. Beside all standard features some special things have been taken over from the already existing DDJ-200 script or added:

- 4 Deck control possible
- Select and load tracks from within the controller
- use pad buttons for
  - hotcues (ok, standard)
  - loops
  - effects and samples
  - beatjump

See
- https://github.com/absorb-it/Pioneer-DDJ-200-Mixxx
- https://mixxx.discourse.group/t/new-mapping-for-pioneer-ddj-200/31164/4
- https://mixxx.discourse.group/t/pioneer-ddj-200-mapping/18259/19

Commit:
Pioneer DDJ-200: fix for hotcue issues, rewrite

fix for issues with hotcue buttons. Complete rewrite, code cleanup

Swiftb0y commented:
Hey there Rene, I appreciate the rewrite with ComponentJS, The resulting implementation is much better than the previous one. However, there is still lots I would like to improve via code review. Would you be okay with spending some additional time getting this into an even better shape?

Swiftb0y commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -69,338 +73,438 @@ DDJ200.shutdown = function() {
+                // song duration in order for the jog wheel to cover the same amount
+                // of time given a constant turning angle.
+                const duration = engine.getValue(group, "duration");
+                const newPos = Math.max(0, oldPos + (this.inValueScale(value) * 0.2 / duration));
```
you can seek into the pre-roll (and there are legit reasons for doing so), so I would remove the `Math.max` here.

Swiftb0y replied:
yeah thats fair I guess. I have an absolute position strip on my controller, so seeking back to the start is easy there...

Swiftb0y commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -69,338 +73,438 @@ DDJ200.shutdown = function() {
+            this.blinkConnection = undefined;
+        };
+        switch (DDJ200.padModes[DDJ200.padModeIndex]) {
+        case "Hotcue":
+            midi.sendShortMsg(0x96, 0x59, 0x00);
+            midi.sendShortMsg(0x96, 0x5A, 0x00);
+            break;
```
use `sendShifted` and `this.send()` (I consider `midi.sendShortMsg` an antipattern in ComponentJS mappings, because it usually means that a component is not used as intended (one component per input/output combination)).

absorb-it replied:
removed with componentContainer implementation 6d70155bfba20eb2064ee0190f1e939e6da31060

Swiftb0y commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -69,338 +73,438 @@ DDJ200.shutdown = function() {
+
+DDJ200.padModes = ["Hotcue", "Loop", "Effect", "Jump"];
+DDJ200.padModeIndex = 0;
+DDJ200.transFxButton = new components.Button({
+    blinkConnection: undefined,
+    input: function(_channel, _control, value, _status, _g) {
+        if (value) { // only if button pressed
+            if (DDJ200.padModeIndex === (DDJ200.padModes.length - 1)) {
+                DDJ200.padModeIndex = 0;
+            } else {
+                if (DDJ200.leftDeck.shiftPressed || DDJ200.rightDeck.shiftPressed) {
+                    DDJ200.padModeIndex = DDJ200.padModes.length - 1;
+                } else {
+                    DDJ200.padModeIndex = (DDJ200.padModeIndex + 1) % (DDJ200.padModes.length - 1);
+                };
+            };
+            DDJ200.leftDeck.pads.forEach(function(e) {
+                e.updateOutKey();
+            });
+            DDJ200.rightDeck.pads.forEach(function(e) {
+                e.updateOutKey();
+            });
+            this.output();
+        };
```
rework with `ComponentContainter`s (see previous comments). Also consider using `script.posMod` and/or a helper like `NS6II.RinBufferView` (feel free to copy-paste).

Swiftb0y reviewed (commented):
This should be plenty of feedback already. I will do my best to help you resolve them along the way.

Commit:
Pioneer DDJ-200: added eslint excemption

absorb-it commented on res/controllers/Pioneer DDJ-200.midi.xml:
```diff
@@ -1,959 +1,903 @@
+        </options>
+      </control>
+      <control>
+        <group>[Channel2]</group>
+        <key>DDJ200.rightDeck.jogWheel.inputWheel</key>
+        <description>Deck 2: Scratch</description>
+        <status>0xB1</status>
+        <midino>0x23</midino>
+        <options>
+          <script-binding />
+        </options>
+      </control>
+      <control>
+        <group>[Channel1]</group>
+        <key>DDJ200.leftDeck.jogWheel.inputWheel</key>
+        <description>Deck 1: Scratch</description>
+        <status>0xB0</status>
+        <midino>0x23</midino>
+        <options>
+          <script-binding />
+        </options>
+      </control>
```
Maybe you skipped one line in JS documentation. 0x23 is ok, 0x29 is used together with Shift and forwarded to overlayed DDJ200.leftDeck.jogWheel.inputSeek

absorb-it committed:
Pioneer DDJ-200: removed volume/gain changes

removed volume and gain changes, they don't belong into the mapping

Co-authored-by: Swiftb0y <12380386+Swiftb0y@users.noreply.github.com>

Commit:
Pioneer DDJ-200: don't override user options

user can choose this in GUI, don't override it

Swiftb0y commented:
> Already used pre-commit, but this changed the code into non-working. 

This should not happen... If that is the case, please push the broken code temporarily so we can take a look at it.

Swiftb0y commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -69,338 +73,438 @@ DDJ200.shutdown = function() {
+    });
+
+    this.rate = new components.Pot({
+        inKey: "rate",
+        invert: 1,
+    });
```
> Slider gives 0x00 at the top, 0x7F at the bottom.

Right, that is because the label on the controller is that "down increases speed" (+ sign on the bottom, - at the top). So always inverting here would be wrong because it mismatches the label on the controller. So by default, you should _not_ invert the value here. 

<img width="256" height="386" alt="image" src="https://github.com/user-attachments/assets/7955921b-d4bb-4835-9b70-dc45f5b8e6fe" />


> "Invert" seems to be required, the suggested GUI option only changes the behavior of the slider at all, but without the invert moving down on the hardware deck moves the slider up in the gui. 

Right, that is because by default the GUI slider has + on the top and - on the bottom, which is inverted compared to your controller. If you want this to match your controller, you need to tick the "down increases speed" option in the GUI. Then the controller will match your onscreen display and its label.

If you feel like this is counter-intuitive (I don't blame you, but you'll need to complain to Pioneer about their questionable tempo slider labeling), you can invert the slider here, but please make it an opt-in option instead of the default. See this section of the wiki how you can add user-friendly options to your mapping. https://github.com/mixxxdj/mixxx/wiki/midi%20scripting#settings-api

absorb-it replied:
It's definitely counter-intuitive even I understand this option much better now... but I'm still not convinced :)

The Hardware Slider on the Deck is now down (+):

without invert:
with DownIncreasesSpeed off: GUI on top, +8
with DownIncreasesSpeed on: GUI on top, -8

with invert:
with DownIncreasesSpeed off: GUI on bottom, -8
with DownIncreasesSpeed on: GUI on bottom, +8

I still think the invert option is right here.

Swiftb0y replied:
yeah, my Numark does that too...

Swiftb0y replied:
> I still think the invert option is right here.

thanks for the table. I think you are indeed correct and I was mistaken. I didn't think "down increases speed" influences the way the `rate` value was interpreted. Keep the `invert` then.

Commit:
Pioneer DDJ-200: shutdown will switch off all LEDs

there is no need for any additional function

absorb-it commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -69,338 +73,438 @@ DDJ200.shutdown = function() {
+            midi.sendShortMsg(0x97 + 2 * i, j, 0x00); // hotcue
+        };
+    };
+};
```
ok, implemented this in c62d61003bbdf358ccd5a91b6a1c62af14a49569 and removed unnecessary LED functions.

Commit:
Pioneer DDJ-200: moved Button initializations into init

initializing buttons before init might result in trouble, moved them
into the init function

absorb-it commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -69,338 +73,438 @@ DDJ200.shutdown = function() {
+
+DDJ200.headMixButton = new components.Button({
+    midi: [0x96, 0x63],
+    key: "headMix",
```
fixed with 43739b8dc733aa6f24f75cec52e1437013d7d673.
Cleanup of ShiftButton handling and ComponentContainer switching as mentioned in other comments still to be done.

absorb-it committed:
Pioneer DDJ-200: use inKey for loading tracks

use inKey instead of separate input function

Co-authored-by: Swiftb0y <12380386+Swiftb0y@users.noreply.github.com>

Commit:
Pioneer DDJ-200: disconnect outKey before change

first disconnect, than change the outKey

absorb-it commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -69,338 +73,438 @@ DDJ200.shutdown = function() {
+                    engine.setValue(this.group, "rateRange", (currentRange>0.9)?0.1:currentRange+0.1);
+                };
+            };
+        },
+    });
+
+    this.playButton = new components.PlayButton({
+        outKey: "play_indicator",
+        shift: function() {
+            this.inKey = "reverse";
+            this.midi = [0x8F + midiChannel, 0x47];
+        },
+        inputFaderStart: function(channel, control, value, status, _g) {
+            this.inKey = "play";
+            this.input(channel, control, value, status, _g);
+            this.inKey = "reverse";
+        },
+        unshift: function() {
+            this.inKey = "play";
+            this.midi = [0x8F + midiChannel, 0x0B];
+        },
+    });
+
+    this.cueButton = new components.CueButton({
+        outKey: "cue_indicator",
+        shift: function() {
+            this.inKey = "cue_gotoandstop";
+            this.midi = [0x8F + midiChannel, 0x48];
+        },
+        inputFaderStop: function(channel, control, value, status, _g) {
+            this.inKey = "cue_set";
+            this.input(channel, control, value, status, _g);
+            this.inKey = "cue_gotoandstop";
+            this.input(channel, control, value, status, _g);
+        },
+        unshift: function() {
+            this.inKey = "cue_default";
+            this.midi = [0x8F + midiChannel, 0x0C];
+        },
+    });
```
could not get default implementation to respect the current deck of the 4-deck selection - maybe I did something wrong with selecting the related deck?

absorb-it replied:
cleanup done with d83e43b7c247ebb421d78e1e25741c83ca714507, but issue persists.

absorb-it replied:
> mhmm. Weird. I'll take a look.

I will have another look tonight - I had read "default implementation" always as using `<Button />` in the midi.xml, but never tried using the `<script-binding/>` without additional settings. Will get back on the topic...

Swiftb0y replied:
With "default implementation" I just meant to use `components.PlayButton({midi:[...]});` (and `CueButton`) without any additional customization. They're already customized to work well out-of-the-box with the user-configurable cue modes.

Commit:
Pioneer DDJ-200: cleanup shift button behaviour

simplified shift button behaviour and added buttons to related decks

Commit:
Pioneer DDJ-200: play and cue btn cleanup

cleanup of play and cue btn implementation

absorb-it commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -69,338 +73,438 @@ DDJ200.shutdown = function() {
+DDJ200.shiftButton = new components.Button({
+    input: function(_channel, _control, value, status, _g) {
+        DDJ200.leftDeck.shiftPressed = value && (status === 0x90);
+        DDJ200.rightDeck.shiftPressed = value && (status === 0x91);
```
fixed with dbaa9bdb2794d79b54148afa4af5f3c63edf95f8

absorb-it commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -69,338 +73,438 @@ DDJ200.shutdown = function() {
+                this[`output${DDJ200.padModes[DDJ200.padModeIndex]}`](value, g, control);
+            },
+            midiNormal: [0x97 + (midiChannel - 1)*2, 0x00 + i],
+            midiShift: [0x98 + (midiChannel - 1)*2, 0x00 + i],
```
fixed with dbaa9bdb2794d79b54148afa4af5f3c63edf95f8

Swiftb0y commented:
Oh, make sure to run pre-commit on this entire branch once, it can cause issues if you've already made some commits and only installed pre-commit afterwards. You can do that by doing `pre-commit run -s 2.5 -o HEAD`

absorb-it commented:
> > eslint changed the
> > var DDJ200 {};
> > into
> > const DDJ200 {};
> > which failed.
> 
> weird. It should not do that with the `// eslint-disable`. I'll take a look.

The  `// eslint-disable` was the solution of the problem. I did not had that line and eslint did a good bunch of formatting changes and just this small replacement of `var`with `const` - which rendered the whole thing broken.

Therefore - nothing to be looked at, but maybe some small comment in the docu why this eslint- check fails might be interesting.

Commit:
Pioneer DDJ-200: use components-js implementations

use component-js implementations where possible

Commit:
Pioneer DDJ-200: rewrote to use componentContainers

Now using componentConatiners similar to Numark NS6II implementation.

Commit:
Pioneer DDJ-200: use all 4 effectunits

change effectunit when changing the Decks

Commit:
Merge branch 'compContainer' into pioneer_ddj-200

cleanup and usage of componentContainers for pad switching

Commit:
Pioneer DDJ-200: fix for jogWheel

was broken with last commit, fixed

Swiftb0y commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -45,123 +45,366 @@ Deck n = 0/1:                                   Midi in         Midi out
+            });
+        } else if (i < 7) {
+            return new DDJ200.Pad(i, deckOffset, {
+                midi: [0x97 + deckOffset, i],
+                input: function(_channel, _control, value, _status, _g) {
+                    const effect = `[EffectRack1_EffectUnit${ script.deckFromGroup(this.group) }_Effect${ i - 3 }]`;
+                    engine.setValue(effect, "enabled", value);
+                    this.send(this.outValueScale(value));
+                },
+                outValueScale: function(_value) {
+                    const effect = `[EffectRack1_EffectUnit${ script.deckFromGroup(this.group) }_Effect${ i - 3 }]`;
+                    return engine.getValue(effect, "loaded")?0x7F:0x00;
+                }
```
```suggestion
                midi: [0x97 + deckOffset, i],
                group: `[EffectRack1_EffectUnit${ script.deckFromGroup(this.group) }_Effect${ i - 3 }]`,
                key: "enabled",
```

absorb-it replied:
Fixed with e4be97c0ae5a347ba59e084c62d8e33c25b63330

Swiftb0y commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -45,123 +45,366 @@ Deck n = 0/1:                                   Midi in         Midi out
+                    this.send(this.outValueScale(value));
+                },
+                trigger: function(_value) {
+                    this.send(0x00);
```
```suggestion
                    this.send(this.off);
```

absorb-it replied:
fixed with e7859cb5761334a2df1912c32753918b014023b8

Swiftb0y commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -45,123 +45,366 @@ Deck n = 0/1:                                   Midi in         Midi out
 */
 
+/*
+    Disabling Demo Mode according to the user manual:
+    1 - Disconnect the USB cable / power supply
+    2 - Hold both the [SHIFT] and [PLAY/PAUSE] buttons and connect USB cable / power supply
+    3 - Use performance pads to change the settings
+        Pad 1 - Demo mode is switched off
+        Pad 2 - Demo mode starts when you don’t use the unit for 1 minute
+        Pad 3 - Demo mode starts when you don’t use the unit for 5 minutes
+        Pad 4 - Demo mode starts when you don’t use the unit for 10 minutes
+        While saving bottom row of Performance Pads is flashing. Wait till it stops.
+    4 - Disconnect the USB cable / power supply
```
Isn't there also a sysex message we can leverage to force the controller out of demo mode?

absorb-it replied:
It's more a comment because I was looking for a while how to change this behavior. I'm have no idea how to find a suitable sysex message, but changing the demo-mode will be persistent and saved in the controller flash - this should imho. not be changed with the mixx software. If somebody likes this, it's in the manual and also here now.

But if you think it doesn't belong here I can remove the comment.

Swiftb0y commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -45,123 +45,366 @@ Deck n = 0/1:                                   Midi in         Midi out
+        engine.setValue("[Channel3]", "volume", 0);
+        engine.setValue("[Channel4]", "volume", 0);
+    };
+    this.shutdown();
```
why call shutdown on startup?

Swiftb0y commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -45,123 +45,366 @@ Deck n = 0/1:                                   Midi in         Midi out
+DDJ200.PadModeContainers.ModeSelector.prototype = new components.ComponentContainer();
+
+DDJ200.init = function() {
+    this.Decks = {
+        left: new DDJ200.Deck([1, 3], 1),
+        right: new DDJ200.Deck([2, 4], 2),
+    };
```
NIT:
`TitleCase` is usually reserved for class names, use `sname_case` instead. Also if you make this a `ComponentContainer` you can replace all the `Object.values(DDJ200.Decks).forEach(deck => ...);` calls with a slightly nicer `DDJ200.Decks.forEachComponent(deck => ...);`


```suggestion
    this.decks = new components.ComponentContainer({
        left: new DDJ200.Deck([1, 3], 1),
        right: new DDJ200.Deck([2, 4], 2),
    });
```

absorb-it replied:
fixed with 42f5342be24c467449df6092cd1b9117b5ef5381

Swiftb0y reviewed (commented):
Thank you, this looks much better already. I have a couple new comments.

Commit:
Pioneer DDJ-200: final cleanup of play and cue buttons

Commit:
Pioneer DDJ-200: cleanup TransFXButton and RingBuffer

Simplified handling of Pad-Mode switching and Buffer

Commit:
Pioneer DDJ-200: indication fix for PadMode change

fix for issue introduced with last commit

Commit:
Pioneer DDJ-200: transFxButton change in midi

transFxButton will call now input, shift will detected later

Commit:
Pioneer DDJ-200: updated Implementation of Loop Pads

Loop Pad indication must be changed on changes in loop size and if loop
gets enabled or disabled. Used multiple connections.

Commit:
Pioneer DDJ-200: updated Implementation of BeatJump Pads

BeatJump indication must be changed on changes in jump size directly on
jump activation. Used multiple connections.

Commit:
Pioneer DDJ-200: simplified Effect Button implementation

implemented simpler Effect Button version - now only the first two
Effect-Racks can be used. Thats probably enough.

Commit:
Pioneer DDJ-200: updated bpmTap buttons

cleanup of bpmTap button implementation

Commit:
Pioneer DDJ-200: updated implementation of headMix

Commit:
Pioneer DDJ-200: cleanup of sync Button

Commit:
Pioneer DDJ-200: layout changes

Commit:
Pioneer DDJ-200: removed obsolete shutdown at the start

Commit:
Pioneer DDJ-200: Changed DDJ200.decks into ComponentContainer

renamed DDJ200.Decks -> DDJ200.decks, configured as ComponentContainer

Commit:
Pioneer DDJ-200: replaced Object loops with forEachComponentContainer

non-recursive forEachComponentContainer required, recursion will be done
by Components if suitable

Commit:
Pioneer DDJ-200: removed DDJ200.Pad

extra class not useful, only used once

Commit:
Pioneer DDJ-200: changed options and description

changed option defaults and fixed description

Commit:
Pioneer DDJ-200: added option to use SuperEQ as Gain

DDJ-200 misses some hardware Gain Knob for the decks. Added some option
to use the Quick Effects Super Knob on the hardware for Gain control.
Default is off - SuperEQ is SuperEQ.

Commit:
Pioneer DDJ-200: fixed EffectButtons and enabled 3/4th Unit

changed effectButtons to toggle buttons and modified shift level to use
3rd and 4th EffectUnit on shift.

Commit:
Pioneer DDJ-200: fixed PadMode Modeselector

removed externally accessible padInstance Containers (this.loopInstance
-> const loopInstance). These led to multiple trigger connections during
startup - too difficult to remove and handle them all cleanly with
setPads.
Implemented Deck-Switching internally, because without externally
connected padInstances component-js Deckswitching is also not working.

Commit:
Pioneer DDJ-200: minor bugfix für blinking LED

unset blinkConnection if it is not used

Commit:
Pioneer DDJ-200: prevent initialization warning

triggers warning message when called withoput deck number

Commit:
Pioneer DDJ-200: cleanup obsolete loop

not required to loop trough all decks, only the one which button was
pressed has changed

Commit:
Pioneer DDJ-200: solve "missing scratch deck" warning

absorb-it committed:
Pioneer DDJ-200: simplified shift

Co-authored-by: Swiftb0y <12380386+Swiftb0y@users.noreply.github.com>

absorb-it committed:
Pioneer DDJ-200: simplified shutdown

Co-authored-by: Swiftb0y <12380386+Swiftb0y@users.noreply.github.com>

Commit:
Pioneer DDJ-200: fixed Shift issue

Wihout external access to the pads (introduced in
dfca85078b13f20b8a58d0cf1c90ceceb4514c9a) lso also shift/unshift and
shutdown not working. Fixed this.

absorb-it commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -69,338 +73,438 @@ DDJ200.shutdown = function() {
+            },
+            output: function(value, g, control) {
+                this[`output${DDJ200.padModes[DDJ200.padModeIndex]}`](value, g, control);
+            },
```
Implemented with 6d70155bfba20eb2064ee0190f1e939e6da31060, finalized with 265a467c18a3f45e7a3ae7f7d558eda83b8ffc7f and  dfca85078b13f20b8a58d0cf1c90ceceb4514c9a. Fixed.

absorb-it commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -1,406 +1,536 @@
+            component.disconnect();
+        });
+        this.choosenPadInstance = newPadInstance;
+        updateSelectorLeds();
```
removed in latest version

absorb-it commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -1,406 +1,536 @@
+                    return (matchingJumpSize) ? this.on : this.off;
+                }
+                return this.off;
+            },
```
The current implementation works both ways, from GUI to pads and the other way around. The pad related to the choosen jump-size will light up if jump is triggered, also if it is triggered from the GUI. Maybe thats not perfect, but the other suitable option would have been to light up all pads continuously (ready for jump)  - didn't liked that.
With you suggested code the button related to the selected jump size in the GUI is lit up constantly, seems also not helpful for any DJ - because than the lit up button has no relation to the pad functions at all... hmmm...

(using "beatjump_size" is required for this indication to work, this won't work with the beatjump_X_forward / beatjump_X_backward controls)

Swiftb0y commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -1,406 +1,594 @@
-DDJ200.switchPadLED = function(deck, pad, enabled) {
-    midi.sendShortMsg(0x97 + 2 * deck, pad - 1, 0x7F * enabled);
-};
+    this.super1 = new components.Pot({
+        group: (engine.getSetting("useSuperAsGain"))?theDeck.currentDeck:`[QuickEffectRack1_${this.currentDeck}]`,
+        inKey: (engine.getSetting("useSuperAsGain"))?"pregain":"super1",
+    });
```
NIT (this is easier to read):

```suggestion
    if (engine.getSetting("useSuperAsGain")) {
        this.super1 = new components.Pot({
            group: theDeck.currentDeck,
            inKey: "pregain",
        });
    } else {
        this.super1 = new components.Pot({
            group: `[QuickEffectRack1_${this.currentDeck}]`;
            inKey: "super1",
        });
    }
```

Swiftb0y commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -1,406 +1,594 @@
+                    notzero->zero   shift       9n  52  hh      [midi in]       OFF=0x00, ON=0x7F (CUE message for fader start)
+    TransitionFX    press                       96  59  hh      [midi in]       OFF=0x00, ON=0x7F
+                    press           shift       96  5A  hh      [midi in]       OFF=0x00, ON=0x7F
+
```
I just went through the DDJ200 manual and also found that there were dedicated "loaded" LEDs you didn't document or seem to have used `0x9F, i, 0x7F` (where `i` is the channel). Would you mind documenting that and maybe also using in the mapping. Simplify create a plain `Component` and define its `outKey` and no `inKey`.

Swiftb0y commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -1,406 +1,594 @@
+        shiftChannel: false,
+        shiftControl: true,
+        shiftOffset: 0x3C,
+        inputFaderStart: function(channel, control, value, status, _g) {
+            this.inKey = "play";
+            this.input(channel, control, value, status, _g);
+        },
+    });
+
+    this.cueButton = new components.CueButton({
+        midi: [0x8F + midiChannel, 0x0C],
+        shiftChannel: false,
+        shiftControl: true,
+        shiftOffset: 0x3C,
+        outKey: "cue_indicator",
+        inputFaderStop: function(channel, control, value, status, _g) {
+            this.inKey = "cue_set";
+            this.input(channel, control, value, status, _g);
+            this.inKey = "cue_gotoandstop";
+            this.input(channel, control, value, status, _g);
+        },
+    });
```
can you elaborate a little on how the fader start stuff works? As far as I can tell its a hardware/firmware feature, right? So if you use Shift plus the channel- or crossfader, it automatically emits a regular `play`/`cue` midi message? Or does it emit a different messages altogether?

Swiftb0y replied:
I see, In that case, instead of switching around properties of the component, it would be simpler if you just created a separate "faderStart" component (per deck I guess?)

absorb-it replied:
Just checked the manual again and Faderstart seams to be better be replaced with "cue_gotoandplay". Implemented this and some comment with 81ac26c5fd31f039f55ad9d6854cfd050aa9eee4

Swiftb0y commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -1,406 +1,594 @@
+    this.decks = new components.ComponentContainer({
+        left: new DDJ200.Deck([1, 3], 1),
+        right: new DDJ200.Deck([2, 4], 2),
+        shift: function() {
+            this.left.padUnit.choosenPadInstance().shift();
+            this.right.padUnit.choosenPadInstance().shift();
+            components.ComponentContainer.prototype.shift();
+        },
+        unshift: function() {
+            (this.left.padUnit.choosenPadInstance()).unshift();
+            this.right.padUnit.choosenPadInstance().unshift();
+            components.ComponentContainer.prototype.unshift();
         }
```
can you elaborate on this? I don't think it does what you want it to.

absorb-it replied:
first, the outer brackets in line 346 are leftovers, I will remove them  - but they are not relevant.

For the code, I added  some direct call to left and right padUnit shift/unshift, because these are functions which will apply shift/unshift to my selected PadContainer. After that the prototype function is called, which applies shift/unshift to all the rest.

This was some result of the changes in dfca85078b13f20b8a58d0cf1c90ceceb4514c9a which made the PadContainers not part of any automated loops from the outside - this gave a lot of troubles with double and leftover connections.

Works as implemented - but maybe it's not ok to call the prototype function directly?

Swiftb0y replied:
Well I should've actually explained my issues with this snippet :sweat_smile: 

1. If you call `this.right.padUnit.choosenPadInstance().unshift();` from this scope here, its still very easy to call `this.right.padUnit.unshift();` leading to the bugs you describe, so instead I would shift this delegation to the `padUnit` instance itself (so it ensures it only shifts the currently active layer. This would make it harder to accidentally call (`un`)`shift` on all layers and would also simplify the code here.
2. Unless I'm deeply mistaken in my JS knowledge, `components.ComponentContainer.prototype.unshift();` calls shift on _a_ global `components.Component` instance (which just does nothing). Since you probably want to call the "parent" function here, you'll need to first access the proper class (no `.prototype`) and second explicitly tell the function you call  `this`, to run on. So `components.ComponentContainer.unshift.call(this);`. You have already encountered this pattern to call the parent constructor (for instance using `components.Deck.call(this, ...);`). ES6 class syntax helps you to avoid these pitfalls by providing convenient shortcuts (such as the `super` keyword), though the underlying mechanisms still exist.

Swiftb0y replied:
> won't work, because shift/unshift are functions of the prototype.

I don't think so? It shouldn't matter in practice though. If a property is not defined on an object, the interpreter automatically walks the prototype chain to see if it is defined there (until the prototype is undefined). But yeah, you probably don't need this anyways because if so, you would've noticed the bug earlier yourself :)

absorb-it replied:
fixed with d6521fa8adbbfcd1592f2604feb38522071f3bc6 - finally it's so simple :)

Swiftb0y commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -1,406 +1,594 @@
-    var vgroup = group.replace("Channel" + deckNo, "Channel" + vDeckNo);
-    engine.setValue(vgroup, "parameter" + eq, val);
-};
+    this.updateDeck = function(currentDeck) {
+        padInstancesBuffer.indexable[0].forEach(function(padInstance) {
+            updateDeckHelper(padInstance, currentDeck);
+        });
+        padInstancesBuffer.indexable[0].forEach(function(padInstance) {
+            updateDeckHelper(padInstance, currentDeck);
+        });
+    };
```
this just updates the first padInstance, but you want to update the current instance, right?

Swiftb0y commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -1,406 +1,594 @@
-    // of time given a constant turning angle.
-    var duration = engine.getValue(group, "duration");
-    var newPos = Math.max(0, oldPos + ((value - 64) * 0.2 / duration));
-
-    var deckNo = script.deckFromGroup(group);
-    var vgroup = "[Channel" + DDJ200.vDeckNo[deckNo] + "]";
-    engine.setValue(vgroup, "playposition", newPos); // Strip search
-};
+        startup = false;
```
any particular reason for this `startup` workaround to exist?

absorb-it replied:
This gives me exactly one controlled connection (engine.connect) to the pads. With dfca85078b13f20b8a58d0cf1c90ceceb4514c9a there are no more automated connections - this startup exception ensures that also the first padContainer elements are connected.

Swiftb0y replied:
but wouldn't it be better to just ensure that `setPads` is only run once (in the class constructor for instance)?

Swiftb0y reviewed (commented):
> Now that more and more things are done, please don't accept the PR too fast, it needs some way more testing with all the code changes involved now. Last code change plus one week seems at least reasonable.

Yes of course. Can you clarify what you expect now? IE are you done with the mapping and want to test now and get it merged or are you susceptible to some more feedback on this already very decent mapping? If the latter is the case, I have some more comments for you.

Commit:
Pioneer DDJ-200: fixed JogWheel

fix for issue with JogWheel and Deck switching

absorb-it commented:
> Can you clarify what you expect now? IE are you done with the mapping and want to test now and get it merged or are you susceptible to some more feedback on this already very decent mapping? If the latter is the case, I have some more comments for you.
I will appreciate all the comments I can get. And will do some testing. I don't expect to get way more feedback if posted in the forum, therefore I am happy to get the feedback here and learn while improving the mapping. Once its "ready" it can be approved, but it takes as long as it takes. :)

absorb-it committed:
Pioneer DDJ-200: fixed mixxx version string

Co-authored-by: Swiftb0y <12380386+Swiftb0y@users.noreply.github.com>

absorb-it committed:
Pioneer DDJ-200: improved code readability

Co-authored-by: Swiftb0y <12380386+Swiftb0y@users.noreply.github.com>

Commit:
Pioneer DDJ-200: completed midi documentation

Commit:
Pioneer DDJ-200: bugfix for code typo introduced in dcd6955af53146b1b

Commit:
Pioneer DDJ-200: bugfix wrong index

beatjump unit was not Deck-Shifted

Commit:
Pioneer DDJ-200: fixed typo

Commit:
Pioneer DDJ-200: moved information about DEMO mode to mixx-manual

Commit:
Pioneer DDJ-200: fixed shift and shutdown

fowarded shift, unshift and shutdown to the choosen padinstance

Commit:
Pioneer DDJ-200: separated faderStart and faderStop

Commit:
Pioneer DDJ-200: bugfix wrong deck selected of 4deck is left

wrong deck might have been selected on pads when 4-deck was changed back
to 2-deck view. Fixed.

Commit:
Pioneer DDJ-200: PadModeContainers cleanup

rearranged and cleaned padModeContainer implementation

Commit:
Pioneer DDJ-200: conversion to classes

implemented ES6 classes

Commit:
Pioneer DDJ-200: cleanup + ES6 for padModeContainers

further cleanup padModeCntainers is now a ES6 class

Commit:
Pioneer DDJ-200: removed keyboard flash effect

was not planned for a release -just a test if LED-ready signal is useful
for deck switching signalization - it isn't

Commit:
Pioneer DDJ-200: centralized setCurrentDeck

absorb-it commented on res/controllers/Pioneer-DDJ-200-scripts.js:
```diff
@@ -1,406 +1,536 @@
+            input: function(_channel, _control, value, _status, _mode) {
+                engine.setValue(this.group, "beatjump_size", this.beatJumpSize);
+                engine.setValue(this.group, this.inKey, value);
+            },
```
using "beatjump_size" is required for this indication to work, this won't work with the beatjump_X_forward / beatjump_X_backward controls, see longer comment below

absorb-it commented:
I will do some in-depth checking during the next week. No major code changes planned, everything seems to be working. Lets see!

Commit:
Pioneer DDJ-200: reset Volume of Deck 3 and 4 on start

if not set, it will be 100% on every start - thats not suitable for
starting with these decks.

Commit:
Pioneer DDJ-200: bugfix to prevent double connections

Commit:
Pioneer DDJ-200: readded group configuration

it's required and was removed in last commit - fixed

Commit:
Pioneer DDJ-200: double connection issue cleaned

Setting outConnect: false and adding group to all instances of
PadModeContainer solved finally the problems with double connections.
They had been introduced by connectPadInstance - connection and trigger
was not required there. Initial connections will be established by
components.componentContainer.applyLayer

absorb-it commented:
> I'm not sure either unfortunately. You can use `console.trace()` to get a stacktrace. That should help you figure out where the double `connect` comes from. https://doc.qt.io/qt-6/qtquick-debugging.html#console-api

after a lot of debugging I found the issue - as usual - some simple mistake. Took me a long time to realize that the initial connection of a ComponentContainer, also for Deck come from applyLayer. Thats why the reconnectComponents call was initially in the Deck initialization, but it does not belong there. Which resulted in the double Play button connections.

Beside I had all this trouble with double ModeSelector padMode conections - same reason. They just needed the information about the group, not, as I did, had to be reconnected with connectPadInstance. Supplying the group information during configuration was even possible with the  "outConnect: false" option - finally all comes together...

Let's see if I'm still happy tomorrow, but think I cracked the nut!

Commit:
Pioneer DDJ-200: ensure that indication lights are off

indication lights might stay on if blinking is stopped during on-cycle.
fixed.

Commit:
Pioneer DDJ-200: improved library handling

added option to choose which shift button to use for library scrolling.
pressing boith shift buttons together will now allow you to change
currently focused pane

Commit:
Pioneer DDJ-200: bugfix for sampler buttons

regression with introduced with 6d195e01 . SamplerButton sets own group,
should not be overridden.

Commit:
Pioneer DDJ-200: added option to change rateRange

added Option to change rateRange/Tempo Range even if 4deck mode is used:
pressing both Shift buttons and SyncButton.
unified detection of bith Shift button presses.

Commit:
Pioneer DDJ-200: use dblClick on Shift Btn to swap decks

improved 4-deck handling: just double-clickung on the shift button
changes the selected deck. Handling of Shift + Beat-Sync now mathces the
controller label in all cases and changes Tempo Slider Range.

Commit:
Pioneer DDJ-200: removed Mapping setting duplication

Commit:
Pioneer DDJ-200: fixed wrong pfl button indication

Commit:
Pioneer DDJ-200: reset 4-deck state if changed via gui

change the current deck selection back to 2-deck even if state was
changed via settings gui.

absorb-it committed:
Pioneer DDJ-200: changed controller name

removed leftover alternative naming, set to default controller name

absorb-it commented:
Hi, sorry for some mess I did trough the github interface, I tried to keep this PR based on 2.6 and synced with main branch instead - had to do a hard reset / force push to get rid of this again. But the github-actions bot was faster and added all possible labels to this PR. Seems like I can't remove them, sorry to anybody who has to clean this. Regards, Rene

absorb-it committed:
Pioneer DDJ-200: shift/unshift new padmodeContainer

When changing the PadMode container with shift, the shift-state was not
taken over to the new Pads. Fixed now.
